src/primitives/fieldvec.rs

```diff
@@ -0,0 +1,380 @@
+//!
+
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::vec::Vec;
+use core::{fmt, iter, mem, ops, slice};
+
+use super::Field;
+use crate::primitives::correction::NO_ALLOC_MAX_LENGTH;
+
+/// A vector of field elements.
+///
+/// Parameterized by the field type `F` which can be anything, but for most methods
+/// to be enabled needs `Default` and `Clone`. (Both are implied by `Field`.)
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub struct FieldVec<F> {
+    inner_a: [F; NO_ALLOC_MAX_LENGTH],
+    len: usize,
+    #[cfg(feature = "alloc")]
+    inner_v: Vec<F>,
+}
+
+impl<F> FieldVec<F> {
+    /// Determines whether the residue is representable, given the current
+    /// compilation context.
+    ///
+    /// For small enough residues (which includes, in particular, bech32 and
+    /// bech32m), will always return true. Otherwise, returns true iff the
+    /// **alloc** feature is turned on.
+    ///
+    /// If you just want to panic when this is false, use `assert_has_data`.
+    #[inline]
+    pub fn has_data(&self) -> bool { self.len <= NO_ALLOC_MAX_LENGTH || cfg!(feature = "alloc") }
+
+    /// Panics if [`Self::has_data`] is false, with an informative panic message.
+    #[inline]
+    pub fn assert_has_data(&self) {
+        assert!(
+            self.has_data(),
+            "checksums of {} characters (more than {}) require the `alloc` feature of `bech32` to be enabled",
+            NO_ALLOC_MAX_LENGTH,
+            self.len,
+        );
+    }
+
+    /// Number of stored field elements
+    #[inline]
+    pub fn len(&self) -> usize { self.len }
+
+    /// Whether the vector is empty
+    #[inline]
+    pub fn is_empty(&self) -> bool { self.len == 0 }
+
+    /// Returns an immutable iterator over the elements in the vector.
+    ///
+    /// # Panics
+    ///
+    /// Panics if [`Self::has_data`] is false.
+    pub fn iter(&self) -> slice::Iter<F> {
+        if self.len > NO_ALLOC_MAX_LENGTH {
+            self.assert_has_data();
+            #[cfg(feature = "alloc")]
+            return self.inner_v[..self.len].iter();
+        }
+        self.inner_a[..self.len].iter()
+    }
+
+    /// Returns a mutable iterator over the elements in the vector.
+    ///
+    /// # Panics
+    ///
+    /// Panics if [`Self::has_data`] is false.
+    pub fn iter_mut(&mut self) -> slice::IterMut<F> {
+        if self.len > NO_ALLOC_MAX_LENGTH {
+            self.assert_has_data();
+            #[cfg(feature = "alloc")]
+            return self.inner_v[..self.len].iter_mut();
+        }
+        self.inner_a[..self.len].iter_mut()
+    }
+}
+
+impl<F: Field> FieldVec<F> {
+    /// Constructor from the powers of an element, from 0 upward.
+    #[inline]
+    pub fn from_powers(elem: F, n: usize) -> Self {
+        iter::successors(Some(F::ONE), |gen| Some(elem.clone() * gen)).take(n + 1).collect()
+    }
+
+    /// Multiply the elements of two vectors together, pointwise.
```

Comment:
Needs panics docs.

Reply:
Sure, I'll add them. Though all these docs are private of course.

Reply:
Should have had two coffees, didn't notice the whole `field` module being private yesterday.